### ISSUE TITLE
Characters not supported on translation

### ISSUE THREAD

**Issue by BanCrash** [Kapiainen/Lauhdutin#107]
Hello, I'm making a translation for the new 3.0.0 version and I find that there is 3 characters that aren't supported. I don't know if could be supported, so just in case I let here to know.

Characters are "ú", "Á" and "ü". Third character ("ü") is rarely used in Spanish, so there shouldn't be much trouble if there isn't support for that.

However "Ú", "á" and "Ü" are supported.

By the way, great job with the new version, it looks great!

**Comment by Kapiainen** [comment]
> Characters are "ú", "Á" and "ü". Third character ("ü") is rarely used in Spanish, so there shouldn't be much trouble if there isn't support for that.
> 
> However "Ú", "á" and "Ü" are supported.

All of those characters are in the lookup table that is being used, but I suspect the ones causing issues are represented in another way than what Lauhdutin currently supports. I had a quick look around and it seems like plenty of characters with accents and other additional parts can be represented in many ways in Unicode. For example [this Stack Overflow thread](https://stackoverflow.com/questions/39588795/utf-encoding-for-%C3%9C-returns-3-bytes-instead-of-the-real-unicode) shows an example of how `Ü` can be defined in two different ways. Maybe that is the issue?

**Comment by BanCrash** [comment]
It were my fault. When I saved the txt file, I didn't change Windows default encoding (ANSI). When I set it to UTF-8 the issue were solved. Sorry for the misunderstanding.

**Comment by Kapiainen** [comment]
Okay, thanks for letting me know.

**Comment by BanCrash** [comment]
@Kapiainen Seems there is another issue. When I reload the skin, the txt turns again into ANSI encoding. I suppose the skin writes on the translation file on every startup?

**Comment by Kapiainen** [comment]
It shouldn't do that. Is the first line in your translation file `version 1` like in the default English translation file?

**Comment by BanCrash** [comment]
I didn't change, so should be the same. Should be changed?

https://github.com/BanCrash/Lauhdutin/blob/patch-9/translations/Spanish.txt#L1

By the way, the first line is not working. I put "button_label_browse	Examinar", but in settings shows Browse:

![image](https://user-images.githubusercontent.com/11770760/36738765-a2741e32-1bde-11e8-8cfe-2b6df4f68dca.png)

I change that line on English.txt to "Browser" instead "Browse" but still no change:

![image](https://user-images.githubusercontent.com/11770760/36738933-fea9fa82-1bde-11e8-91e8-aed3dd4d9504.png)

**Comment by Kapiainen** [comment]
I found the cause of that and it will be fixed in the next release. Thanks for letting me know.

**Comment by BanCrash** [comment]
After using the skin, I see that translation file are changed. Maybe it's related to this issue?

![image](https://user-images.githubusercontent.com/11770760/36739944-7641131c-1be1-11e8-8119-5b3edfdc83c0.png)

Perfect, thanks you!

**Comment by Kapiainen** [comment]
The localization system writes to the active translation file under two circumstances:

1) The translation file's version is less than the localization system's version number.

2) The translation file is missing a key-value pair that the localization system is asked for by some part of the skin. The system just writes the key to the end of the file along with `TRANSLATION_MISSING` (or the default value, if the language is set to English).

Are you also sure that the text editor that you are using always saves the file as UTF-8?

**Comment by BanCrash** [comment]
I found the second option on both translations (english default and spanish).

![image](https://user-images.githubusercontent.com/11770760/36741138-a60e7280-1be4-11e8-86f6-945ba9f0a8e9.png)

But I see that line on standar file too:

https://github.com/Kapiainen/Lauhdutin/blob/development-v3/translations/English.txt#L84

> Are you also sure that the text editor that you are using always saves the file as UTF-8?

I think yes, because after saving it as UTF-8 I can see settings properly before the next startup of the skin:

Startup:

![image](https://user-images.githubusercontent.com/11770760/36741396-4bd02466-1be5-11e8-99c7-98cc6cc300ba.png)

After saving translation on UTF-8 and reloading settings.ini:

![image](https://user-images.githubusercontent.com/11770760/36741434-64b248c4-1be5-11e8-8184-269cc4e4e0b8.png)

**Comment by Kapiainen** [comment]
> But I see that line on standar file too:
> 
> https://github.com/Kapiainen/Lauhdutin/blob/development-v3/translations/English.txt#L84

That was a mistake on my part in the MoonScript source. Fixed in the next release.

**Comment by BanCrash** [comment]
Everytime I reload the skin, the translation file is modificated. Before starting:

![image](https://user-images.githubusercontent.com/11770760/36743031-812a7a72-1be9-11e8-8a15-87d1915bd693.png)

After starting:

![image](https://user-images.githubusercontent.com/11770760/36743098-b4dc230c-1be9-11e8-8373-cb249fd197d3.png)

However that doesn't happen on english.txt. 

So I create another txt, with english default but nammed Test.txt. First, untouched:
Before:
![image](https://user-images.githubusercontent.com/11770760/36743392-7d169cf8-1bea-11e8-9ed6-f3705394effc.png)
After that, I unload the skin, wait 3 minutes more or less and load again:
![image](https://user-images.githubusercontent.com/11770760/36743419-923579c4-1bea-11e8-8bd5-d1b8f4c2e2d9.png)

Second test, touched but with default letters (I write "Saveeed" instead "Save"):
Before:
![image](https://user-images.githubusercontent.com/11770760/36743483-b341df9a-1bea-11e8-95b9-938974fad4a7.png)
After:
![image](https://user-images.githubusercontent.com/11770760/36743625-fee7b12c-1bea-11e8-8d38-d114cec7c83d.png)

So seems that when it is a change from the default, the txt is modified. Could be that modification when the problem arises?

**Comment by BanCrash** [comment]
Yes, I think that is the problem. I had the translation file opened in notepad++, with UTF8 encoding selected, and saved. After loading the skin, notepad++ advise me that the file is modified. I reload the file and encoding get back to ANSI.

**Comment by Kapiainen** [comment]
I'm not having any luck in reproducing the issue. I tried with a renamed copy of the default English translation and your Spanish translation. The skin wouldn't write to the file every time that the skin was loaded and unloaded. I also tried deleting a line to force the skin to write the missing key-value pair, which it did, and the encoding remained as UTF-8 (according to Sublime Text 3).

**Comment by BanCrash** [comment]
That's strange. I tried with last beta and with last commit and the issue it's the same in both cases. First time I see the settings correct, but after reloading the skin the problem arises again.

I will make a test with other computer to see if the problem arises there too.

**Comment by BanCrash** [comment]
In the other computer I have the exact problem. First I see all good but when I reload the skin settings appears bad again.

Did you check if with Spanish.txt, in the second reload settings appears correct? For example, the second line of the first tab of settings should be "Número de filas", and not "Nmero de filas".

**Comment by BanCrash** [comment]
I set to only-read spanish translation to try to avoid encoding conversion and this is what I get:
![image](https://user-images.githubusercontent.com/11770760/36755428-7c57a844-1c0c-11e8-9ef1-3c07bf4449cb.png)

But if I make the same with English.txt, nothing happens. All works correctly.

The changes made are:

- The /n converts into a real blank space, as I said before:

![image](https://user-images.githubusercontent.com/11770760/36756052-47cbe188-1c0e-11e8-963c-583297ac129f.png)

- Lines change its place. Before, last line was for initializing, now is for stopping bangs:

![image](https://user-images.githubusercontent.com/11770760/36756280-db053170-1c0e-11e8-880d-7a140dd63db4.png)

Right now I don't see more changes. It is very strange...

**Comment by Kapiainen** [comment]
I updated the localization system, which should now properly escape newlines when writing to the translation file via the class' save method. I also added log messages to the class, which should help reveal what is going on if you enable logging in the skin's settings. There are separate messages for when the entire translation file is written (the class' save method) and when a missing translation is appended to the translation file (the class' get method).

**Comment by BanCrash** [comment]
This is what I get I think is relevant to the matter with the last commit:

![image](https://user-images.githubusercontent.com/11770760/36780687-f6e17b1e-1c73-11e8-8e0c-c258172d57a6.png)

Full log:

![image](https://user-images.githubusercontent.com/11770760/36780736-25b69aa0-1c74-11e8-9386-658896f9855f.png)

**Comment by Kapiainen** [comment]
Odd. I only get the `Loading translation file` regardless of which translation file I use (default, renamed default, or your Spanish translation).

**Comment by BanCrash** [comment]
https://github.com/BanCrash/Lauhdutin/releases/tag/3.0.0Test

I upload a zip file with all the files I used. Please check if with those files doesn't happen to you this issue either.

**Comment by BanCrash** [comment]
I solve the issue doing this:

![image](https://user-images.githubusercontent.com/11770760/36796102-0c29d152-1ca5-11e8-9ed1-457546888490.png)

I also try to remove a line on the translation and the skin write tha line with MISSING_TRANSLATION, but in UTF-8, not in ANSI.

**Comment by BanCrash** [comment]
After that modification the file is still on UTF-8 on every startup. And in settings, I can see all characters properly. However, game.ini still not work properly.

TXT:
![image](https://user-images.githubusercontent.com/11770760/36976144-debd43c6-207c-11e8-93d8-90462ea72727.png)

Game.ini:
![image](https://user-images.githubusercontent.com/11770760/36976172-ee71fe6a-207c-11e8-8a93-0c023a56a899.png)

(ú, Á and ü are not showing).

Same test on settings:

TXT:
![image](https://user-images.githubusercontent.com/11770760/36976212-0fabaafe-207d-11e8-9583-26b363fbf100.png)

Settings.ini:
![image](https://user-images.githubusercontent.com/11770760/36976222-1a0952c6-207d-11e8-9f24-35f04740a5f7.png)

**Comment by BanCrash** [comment]
After more testing on this matter (changing 3 random translation to adding them "Áú"), I see that in Game.ini, Filter.ini and Sort.ini are not working those characters either. However, in settings.ini and search.ini are working.

**Comment by Kapiainen** [comment]
Sorry it has taken so long. I've been banging my head against the wall with this for a while now, but I was able to reproduce the issue and fix it at least on my computer with [this commit](https://github.com/Kapiainen/Lauhdutin/commit/f535551fa7b517856c0dc0b311ee7cd992ef475c). Could you test it as well? You'll need to grab the latest version of the [development-v3](https://github.com/Kapiainen/Lauhdutin/tree/development-v3) branch since the update is not included in the latest public beta.

**Comment by Kapiainen** [comment]
I just pushed a new commit that should fix the issue with e.g. `™` and `®` characters and behave like it did before.

**Comment by BanCrash** [comment]
Perfect! So this seems to be finally solved! Thanks again for your great work!!

**Comment by Kapiainen** [comment]
Thanks for your help and patience in troubleshooting and testing.
